src/lib.rs

```diff
@@ -2489,6 +2562,12 @@ impl Connection {
             aead,
         )?;
 
+        // Once frames have been serialized they are passed to the Recovery
+        // module which manages retransmission. However, some frames are not
+        // retransmittable and storing them is a waste of resources.
+        // So drop them here.
+        frames.retain(|f| f.retransmittable());
```

Comment:
@xanathar you added this. Do the alternative sound reasonable?

Reply:
I double-checked the implementation of the recovery and retransmission parts, and setting the vec of the datagram frame to 0 like below, seems to be enough. Haven't tested yet though.

```rust
    // In frame.rs
    pub fn shrink_for_retransmission(&mut self) {
        match &mut self {
            Frame::Datagram { v } => {
                v.clear();
                v.shrink_to_fit();
            },
            _ => (),
        }
    }

    // in lib.rs, this or a for loop doing the same
   frames.iter_mut().for_each(|f| f.shrink_for_retransmission());
```

Reply:
I took @xanathar's snippet, tweaked it slightly, added a test and made https://github.com/cloudflare/quiche/pull/595/commits/a2b87e32639aca3ef6cce2e5ec5e191c0823ee86

in `lib.rs` rather than do a second loop for shedding, I've just put it inside the `for frame in &frames` loop.